### PR TITLE
Wip cleanup probplot

### PR DESCRIPTION
**Pull request by phobson**
Will squash commits and submit PR against `statsmodels/master` once this is in a better state

TODO:
- serious testing (in Travis and a local machine with MPL)
- `step` plot option per JP's examples and comments 

NBViewer link to show current state (will be updated):
http://nbviewer.ipython.org/gist/phobson/8304312

### CODE REVIEW

File: statsmodels/graphics/gofplots.py [hunk 1]
```diff
@@ -127,34 +127,59 @@ def __init__(self, data, dist=stats.norm, fit=False,
...
-            self.dist = dist
-            self.loc = loc
-            self.scale = scale
-
         # propertes
         self._cache = resettable_cache()
 
     @cache_readonly
+    def dist(self):
+        if self.fit and len(self.fit_params) >2:
```
Comment (josef-pkt) [inline statsmodels/graphics/gofplots.py]: The if sequence doesn't include the case when only loc and scale are estimated. 

I think we can drop the `len(self.fit_params) >2` in the condition

Comment (josef-pkt) [inline statsmodels/graphics/gofplots.py]: `*self.fit_params[:-2]` will be just empty

Comment (phobson) [inline statsmodels/graphics/gofplots.py]: @josef-pkt `self.loc` comes from the `loc` property that includes that logic, I believe. See line 169

File: statsmodels/graphics/gofplots.py [hunk 2]
```diff
@@ -127,34 +127,59 @@ def __init__(self, data, dist=stats.norm, fit=False,
...
+            return self._dist(*self.fit_params[:-2], loc=self.loc,
+                              scale=self.scale)
+        elif len(self._distargs) > 0:
+            return self._dist(*self._distargs, loc=self.loc,
+                              scale=self.scale)
```
Comment (josef-pkt) [inline statsmodels/graphics/gofplots.py]: This also might not catch the case with only fixed loc and scale.

We need the information whether self._dist should be called with loc and scale (this case) or without (next case).
I think we don't have that information yet.

IIUC:
test case:
1) `_dist` is stats.norm, and `self.loc=2, self.scale=2`
2) `_dist` is frozen stats.norm(loc=2, scale=2) and we should return self._dist in this case

File: statsmodels/graphics/gofplots.py [hunk 3]
```diff
@@ -127,34 +127,59 @@ def __init__(self, data, dist=stats.norm, fit=False,
...
+    def fit(self):
+        return self._fit
+    @fit.setter
+    def fit(self, value):
+        self._fit = value
+
+    @cache_readonly
+    def fit_params(self):
+        if self.fit:
+            return self.dist.fit(self.data)
```
Comment (josef-pkt) [inline statsmodels/graphics/gofplots.py:155]: this looks recursive with `dist` ?
maybe it should be `self._dist.fit(self.data)`

**Comment by josef-pkt** [comment]
I think we could merge `fit_params` and `list(distargs) + [loc, scale]` into dist_params, and use it like the current fit_params, instead of treating loc and scale separately.

overall looks good, I need to try out some examples and check the `fit` story again

one suggestion: I would prefer `plot_options` to `plot_kwargs` as a name
(I always think kwargs is something unappetizing.)

**Comment by phobson** [comment]
Sounds good. I'll pick this back up in the evening. --
Paul Hobson
Sorry if this is unintelligible. I'm on my phone.

File: statsmodels/graphics/gofplots.py (hunk 1 above)
Comment (josef-pkt) [inline statsmodels/graphics/gofplots.py]: yes, I see that, however this case is missing (and can be combined with this)

```
if self.fit and len(self.fit_params) == 2:
    return self._dist(loc=self.loc, scale=self.scale)
```

example normal distribution with estimated loc and scale

File: statsmodels/graphics/gofplots.py (hunk 2 above)
Comment (phobson) [inline statsmodels/graphics/gofplots.py]: @josef-pkt if the distribution is frozen, `self.fit` _should_ be `False`. We can rely on that or check to see if `self._dest` has a `name` attribute (frozen distributions don't appear to have `names`). Or is there a better way?

Comment (josef-pkt) [inline statsmodels/graphics/gofplots.py]: Yesterday a ran into this with an example and the exception was that frozen doesn't have a fit method.
That would be a safer way to check. 
another possibility is to let the user explicitly encode it,  for example `distargs=???` `distargs='no'`

Comment (josef-pkt) [inline statsmodels/graphics/gofplots.py]: I'm still thinking about how to best infer this, maybe checking for hasattr(dist, 'numargs') would work better. That's a required attribute for all distributions.
Either way would work for scipy.stats.distributions, but I need to check whether it also works for custom/user distributions. (I'm pretty sure it does and can be checked later if we run into a case where it doesn't work)

**Pull request closed without merging**